### PR TITLE
Initial uv setup for Feluda and image_vec_rep_resnet operator

### PR DESCRIPTION
**Pull request by plon-Susk7**
This PR is associated with issue #430 .

1. Created a new pyproject.toml file specially catered for uv at root folder. The dependencies in this file are meant to run Feluda only and not operators.

2. The template to create this skeleton file is
```
$ uv init --lib feluda
```
By default feluda creates following directory structure on initialization
```
feluda
├── pyproject.toml
├── README.md
├── uv.lock
└── src
    └── feluda
        └── feluda.py
```
but we have made changes to pyproject.toml file to accept a directory structure src/core instead of src/feluda.

3. Create and enter virtual env by
```
$ uv venv
$ source .venv/bin/activate
```
4. Feluda can be made a package using following command
```
$ uv build
```
This will create a dist folder which will contain .whl file which can be installed using 
```
$ uv pip install .
```

5. Similarly other associated packages like operators can be created inside operators folder by running similar initialization command as in step 2 by providing operators name. This will automatically create association in root pyproject.toml file as 
```
[tool.uv.workspace]
members = ["operators/image_vec_rep_resnet"]
```
More about this is mentioned [here](https://docs.astral.sh/uv/concepts/workspaces/#workspace-sources). The main problem with this approach is that it is not suited for cases in which members have conflicting requirements.

6. We can build package for this operator by same command as given in point 4 after moving to specific operator's directory.

7. Same command as 4 to install the package from .whl file.

8. Created a new directory called package_testing to test our library. The code is same as mentioned by @dennyabrain in #430 .

### CODE REVIEW

**Comment by plon-Susk7** [comment]
There was pyproject.toml file already so I have temporarily renamed it to pyproject1.toml.

**Comment by dennyabrain** [comment]
@plon-Susk7 This might be a naive question but would it be possible to build wheels for feluda and image_vec_rep_resnet operator now ?

What I'm thinking is that to fully test isolation, what I'd like to do now is git pull this PR. build wheels for both feluda and the image_vec_rep_resnet operator. Then I'll create a new directory in a completely independent location and copy my python script, config.yml, images and run 
```
pip install path/to/feluda.whl
pip install path/to/image_vec_rep_resnet.whl
python script.py
```

**Comment by plon-Susk7** [comment]
> @plon-Susk7 This might be a naive question but would it be possible to build wheels for feluda and image_vec_rep_resnet operator now ?
> 
> What I'm thinking is that to truly test isolation, what I'd like to do now is git pull this PR. build wheels for both feluda and the image_vec_rep_resnet operator. Then I'll create a new directory in a completely independent location and copy my python script, config.yml, images and run
> 
> ```
> pip install path/to/feluda.whl
> pip install path/to/image_vec_rep_resnet.whl
> python script.py
> ```

@dennyabrain yes, this should work fine. Here's colab link https://colab.research.google.com/drive/1us_dvUZpC4hdVTXnnyGVkE7IvXNAKNfP?usp=sharing

**Comment by dennyabrain** [comment]
Oh wow! This is very exciting to see.
This gives me a lot of confidence to start refactoring operators out of the src folder and start publishing packages :)

I'll let @aatmanvaidya review and merge the code.

File: package_testing/test.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
+from core.feluda import Feluda
```
Comment (aatmanvaidya) [inline package_testing/test.py:1]: hi @plon-Susk7 wanted to ask a question, can we move the entire testing code out of feluda? and just run it in an isolated environment like google colab etc?

don't know what the ideal protocol is, hence asking
also what if we write new tests for such use cases? like using a config to init Feluda class and use the opeartor.

File: package_testing/test.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+from sklearn.metrics.pairwise import cosine_similarity
+from  core.models.media_factory import ImageFactory
+
+feluda = Feluda('config.yaml')
+feluda.setup()
+
+feluda.setup()
```
Comment (aatmanvaidya) [inline package_testing/test.py:8]: there is a repetition of `feluda.setup()`

File: test.py [hunk 3]
```diff
@@ -0,0 +1,5 @@
+from core.feluda import Feluda
```
Comment (aatmanvaidya) [inline test.py:1]: is this file needed? or shall we remove it

File: config.yaml [hunk 4]
```diff
@@ -0,0 +1,6 @@
+operators:
```
Comment (aatmanvaidya) [inline config.yaml:1]: does this `config.yaml` have to be at `src`, are we using this somewhere?

File: package_testing/test.py (hunk 1 above)
Comment (plon-Susk7) [reply package_testing/test.py:1]: hey @aatmanvaidya , here's an example that I ran in google colab https://colab.research.google.com/drive/1us_dvUZpC4hdVTXnnyGVkE7IvXNAKNfP?usp=sharing

File: package_testing/test.py (hunk 2 above)
Comment (plon-Susk7) [reply package_testing/test.py:8]: I'll make changes right away

File: test.py (hunk 3 above)
Comment (plon-Susk7) [reply test.py:1]: test.py is just for testing if the packages are running fine. We could remove it and add link to google colab that I mentioned above.

File: package_testing/test.py (hunk 1 above)
Comment (aatmanvaidya) [reply package_testing/test.py:1]: hi @plon-Susk7 I did take a look at the colab, the code there looks great!!
I wanted to ask if the test file in the code is necessary then? if yes, we should write like proper tests for it where we assert values etc -- this makes it more robust and easy to maintain

File: config.yaml (hunk 4 above)
Comment (plon-Susk7) [reply config.yaml:1]: The config.yaml needs to be at the directory where we run our script since we are passing it to Feluda class instance.

File: operators/image_vec_rep_resnet/pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,21 @@
+[project]
```
Comment (aatmanvaidya) [inline operators/image_vec_rep_resnet/pyproject.toml:1]: a questions about this file

if I understand correctly, each operator will require a `pyproejct.toml` file right?
and current we have a folder at source of Feluda called `operators`, inside which we have `operators/image_vec_rep_resnet` and inside it there is a `.toml` file and a `src/image_vec_rep_resnet/` folder -- will we have to maintain this structure for each operator? or can we have all operators at one place?

Comment (plon-Susk7) [reply operators/image_vec_rep_resnet/pyproject.toml:1]: We have to maintain that structure for each operator since we are releasing each package separately. Each operator also has different dependencies so it makes sense to make different .toml for different operators.

File: test.py (hunk 3 above)
Comment (aatmanvaidya) [reply test.py:1]: yes but there is a `test.py` file in the `package_testing` folder and this one is at the root level, and this file just has the code to init Feluda class, hence wanted to clarity if I missed something or not

Comment (plon-Susk7) [reply test.py]: oh right, I meant to just test if the package are running fine. Sure we can write dedicated tests for it.

File: package_testing/test.py (hunk 1 above)
Comment (plon-Susk7) [reply package_testing/test.py:1]: I agree @aatmanvaidya. We can remove the the current file and instead add proper tests.

File: operators/image_vec_rep_resnet/pyproject.toml (hunk 5 above)
Comment (aatmanvaidya) [reply operators/image_vec_rep_resnet/pyproject.toml:1]: I see understood!

so a .toml file will fact as a requirements file for each operator and will allow the user to install each operator separately perfect!

File: package_testing/test.py (hunk 1 above)
Comment (aatmanvaidya) [reply package_testing/test.py:1]: great, so to just test the operator, we just need a unit test that extracts the embedding on an image
I think such a test for the operator might already exist, 

we should write a test that init's Feluda class and then uses the operator, extracts the embedding on an image (you can just use one of the links I had provided earlier), and asserts for the vector length

File: pyproject_1.toml [hunk 6]
```diff
@@ -0,0 +1,11 @@
+[tool.semantic_release]
```
Comment (aatmanvaidya) [inline pyproject_1.toml:1]: also, do you think there is a way to merge both the `pyproject.toml` files, or will there be some conflicts?
because for `uv` specific things, I see a heading of `[tool.uv.workspace]` etc, so can both the files be merged into one?

Comment (plon-Susk7) [reply pyproject_1.toml:1]: All the dependencies are captured in uv.lock file. Maybe that solves our problem?

File: package_testing/test.py (hunk 1 above)
Comment (plon-Susk7) [reply package_testing/test.py]: @aatmanvaidya , I have added the test case.

Comment (plon-Susk7) [reply package_testing/test.py]: After initializing the virtual environment move to package_testing directory and run following command.
```cmd
uv run -m unittest test_image_vec_rep_resnet.py
```

Comment (aatmanvaidya) [reply package_testing/test.py]: it looks great, one small thing, can you rename the folder from `package_testing` to `tests`. And the function you have written for `test_sample_image_from_disk()` is perfect, but I think we should not populate the codebase right now with more sample images, we already have enough of them, so you can skip the disk() function?

File: pyproject_1.toml (hunk 6 above)
Comment (aatmanvaidya) [reply pyproject_1.toml:1]: I think it should, can you do some reading around it? 
hopefully we should be able to merge `pyproject_1.toml` and `pyproject.toml` into a single file

File: package_testing/test.py (hunk 1 above)
Comment (plon-Susk7) [reply package_testing/test.py]: done

Comment (aatmanvaidya) [reply package_testing/test.py]: you don't have to delete that function - you can just add a `@skip` above the function
```py
from unittest.case import skip

@skip
test_sample_image_from_disk():
```
you can see this for example - [code file](https://github.com/tattle-made/feluda/blob/690d48a63a855cfa8675fd1099f57e8a6aacd8db/src/core/operators/test_vid_vec_rep_clip.py#L19-L21)

also, the `package_testing` folder is still there, can you also delete it?

File: pyproject_1.toml (hunk 6 above)
Comment (aatmanvaidya) [reply pyproject_1.toml:1]: do you think we should keep this change as the scope of this PR? happy to take it later as well

**Comment by aatmanvaidya** [comment]
@plon-Susk7 can you also give me detailed instructions on how do I run the PR locally. I read your top comment (PR description)
I am guessing I won't have to run 
```sh
uv init --lib feluda
```

I will just have to create a venv right?
```sh
uv venv
source .venv/bin/activate
```
does this override the existing `.venv` folder?

and then run `$ uv build`, this gives me a wheel file for feluda pacakge right? if yes, how do I generate wheel files for the operators?

**Comment by plon-Susk7** [comment]
> @plon-Susk7 can you also give me detailed instructions on how do I run the PR locally. I read your top comment (PR description) I am guessing I won't have to run
> 
> ```shell
> uv init --lib feluda
> ```
> 
> I will just have to create a venv right?
> 
> ```shell
> uv venv
> source .venv/bin/activate
> ```
> 
> does this override the existing `.venv` folder?
> 
> and then run `$ uv build`, this gives me a wheel file for feluda pacakge right? if yes, how do I generate wheel files for the operators?

Yes this does override venv folder unless we specify some other name for virtual environment. You don't have to initialise anything here since the structure is already there. Running build command on root directory will create wheel for feluda wheras moving to operators/{name of operator} and running build command will create wheel for the operator. Both can be found .dist folder.

File: pyproject_1.toml (hunk 6 above)
Comment (plon-Susk7) [reply pyproject_1.toml:1]: Yeah I think we should leave it as it is for this PR as a reference of our working project. We can take this up in a separate issue.

File: tests/test_image_vec_rep_resnet.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+        # Intializing Feluda
+        cls.feluda = Feluda('image_vec_rep_resnet-config.yaml')
+        cls.feluda.setup()
+
+    @classmethod
+    def tearDownClass(cls):
+        # Clean up resources if needed
+        pass
+
+    @skip
```
Comment (aatmanvaidya) [inline tests/test_image_vec_rep_resnet.py:18]: hi @plon-Susk7 one small last change is needed, you forgot to import skip
```py
from unittest.case import skip
```
the test will fail otherwise

Comment (plon-Susk7) [reply tests/test_image_vec_rep_resnet.py:18]: Really sorry for overlooking this. I've made the changes.

Comment (aatmanvaidya) [reply tests/test_image_vec_rep_resnet.py:18]: no no all good, things look good now!

will test once locally and merge the PR!
